### PR TITLE
Add notebook controller culling

### PR DESCRIPTION
**Pull request by maroroman**
This PR implements this PR: https://github.com/opendatahub-io/odh-manifests/pull/581 into the cluster settings interface.

No changes from UI side should be neccessary. The backend does not set the `IDLENESS_CHECK_PERIOD` and ignores pvc_size values if notebook controller is enabled. It also ignores jupyterhub settings (pvc_size, culler_timeout) and does not patch them if notebook controller is enabled in the DashboardConfig CR.

I added an exception for const because eslint was firing errors at my tuple, let me know if I should revert that.

Resolves:https://github.com/opendatahub-io/odh-dashboard/issues/353

### CODE REVIEW

**Review by samuelvl — APPROVED** [review]

**Review by DaoDaoNoCode — APPROVED** [review]
lgtm

**Comment by samuelvl** [comment]
/hold

Do not merge until https://github.com/opendatahub-io/odh-dashboard/pull/298 is merged before

**Review by samuelvl — CHANGES_REQUESTED** [review]
When I click on "Save changes" nothing happens:

image

I can see this error in the dashboard logs:

```
{"level":50,"time":1658313323526,"pid":68,"hostname":"odh-dashboard-d5b58f889-l6qnc","reqId":"req-7g","err":{"type":"FastifyError","message":"Promise may not be fulfilled with 'undefined' when statusCode is not 204","stack":"FastifyError: Promise may not be fulfilled with 'undefined' when statusCode is not 204\n    at /opt/app-root/src/backend/node_modules/fastify/lib/wrapThenable.js:30:30\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","name":"FastifyError","code":"FST_ERR_PROMISE_NOT_FULFILLED","statusCode":500},"msg":"Promise may not be fulfilled with 'undefined' when statusCode is not 204"}
```

**Comment by maroroman** [comment]
Latest changes should address your comment @samuelvl

**Comment by andrewballantyne** [comment]
What is our ticket backing this @maroroman -- https://github.com/opendatahub-io/odh-manifests/pull/581 is closed.

We need a ticket to track the work for this PR. Either we can reopen 581 or we need to log a new one and update your description.

**Comment by maroroman** [comment]
> What is our ticket backing this @maroroman -- [opendatahub-io/odh-manifests#581](https://github.com/opendatahub-io/odh-manifests/pull/581) is closed.
> 
> We need a ticket to track the work for this PR. Either we can reopen 581 or we need to log a new one and update your description.

Linked this PR to a new issue https://github.com/opendatahub-io/odh-dashboard/issues/353

**Review by samuelvl — CHANGES_REQUESTED** [review]
- The `notebook-controller-culler-config` configmap shouldn't be automatically created when I open the `Cluster settings` tab:

image

"Do not stop idle notebooks" should be chosen by default until I select "Stop idle notebooks after".

- When I select a different culling configuration I get the following error:

image

- The `notebook-controller-culler-config` configmap is not deleted when "Do not stop idle notebooks" is selected and I save changes:

image


``` $ oc get cm notebook-controller-culler-config                  
NAME                                DATA   AGE
notebook-controller-culler-config   3      11m
```

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 1]
```diff
@@ -81,20 +115,66 @@ export const getClusterSettings = async (
...
   }
+
   return clusterSettings;
 };
+
+const readJupyterhubCfg = async (fastify: KubeFastifyInstance): Promise<[number, number]> => {
+  const jupyterhubCfgResponse = await fastify.kube.coreV1Api.readNamespacedConfigMap(
+    jupyterhubCfg,
+    fastify.kube.namespace,
+  );
```
Comment (andrewballantyne) [inline backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:176]: Recommend you do a `.then` on this and normalize your data away from the kube library's response.

Eg something like:
```
.then((response) => response.body)
```

That way your variable `jupyterhubCfgResponse` can just be a `ConfigMap`.

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 2]
```diff
@@ -81,20 +115,66 @@ export const getClusterSettings = async (
...
+  const pvcSize = Number(jupyterhubCfgResponse.body.data.singleuser_pvc_size.replace('Gi', ''));
```
Comment (andrewballantyne) [inline backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:177]: Same issue here, what if it is less than a gig?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: backend/.eslintrc [hunk 3]
```diff
@@ -40,6 +40,7 @@
     "@typescript-eslint/no-var-requires": "off",
     "@typescript-eslint/no-empty-function": "warn",
     "@typescript-eslint/no-inferrable-types": "error",
+    "prefer-const": "warn",
```
[the remark this answers is not in the document]
Comment (maroroman) [reply backend/.eslintrc:43]: There was an issue I was having with a tuple, but it was just new syntax for me, removed the warn.

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 4]
```diff
@@ -39,30 +42,61 @@ export const updateClusterSettings = async (
...
+            },
+          },
+        );
+      }
+    }
+    if (dashConfig.spec.notebookController.enabled) {
+      await rolloutDeployment(fastify, namespace, 'notebook-controller-deployment');
+    } else {
+      const jupyterhubCM = await coreV1Api.readNamespacedConfigMap(jupyterhubCfg, namespace);
       if (jupyterhubCM.body.data.singleuser_pvc_size.replace('Gi', '') !== query.pvcSize) {
```
[the remark this answers is not in the document]
Comment (maroroman) [reply backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:94]: The UI doesn't seem to allow any size lower than 1Gi so I assumed Gi here

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:94]: I'm not entirely sure if that's the case. We'll be adding configurations in to allow for custom notebooks, I'd imagine they have the power to do whatever they want. But we can go with this for now if it's not possible to adjust it any lower.

I know this one is on JH config map, so maybe it's not the same restriction for KFNBC?

[the remark this answers is not in the document]
Comment (erwangranger) [reply backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:94]: this drives the PVC size. On AWS, these PVCs are backed by EBS volumes. (sorry for the acronym soup). 
An EBS volume cannot be less that 1 GiB.

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:94]: Well, that sounds pretty safe to assume Gi then. I appreciate you jumping in here @erwangranger -- seemed like a logical fallacy to assume gig, glad there is something preventing us from worrying about it.

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 5]
```diff
@@ -39,30 +42,61 @@ export const updateClusterSettings = async (
...
+          undefined,
+          undefined,
+          {
+            headers: {
+              'Content-Type': 'application/merge-patch+json',
+            },
+          },
+        );
+      } else {
+        await coreV1Api.patchNamespacedConfigMap(
```
[the remark this answers is not in the document]
Comment (maroroman) [reply backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts]: Wrapped the patch into a function.

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 6]
```diff
@@ -20,54 +23,72 @@ export const updateClusterSettings = async (
...
-        },
-      );
     }
     return { success: true, error: null };
   } catch (e) {
     if (e.response?.statusCode !== 404) {
-      fastify.log.error('Setting cluster settings error: ' + e.toString());
+      console.log(e.response);
+      console.log(e.response.body);
+      console.log(e.response.body.message);
```
Comment (andrewballantyne) [inline backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:90]: Please remove debug code

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Review by samuelvl — APPROVED** [review]
It works great! 

/lgtm

**Review by andrewballantyne — APPROVED** [review]

**Pull request merged**